TestManager_FakeInput_OutputChange test is flaky

## Flaky Test

This has only been observed on Windows so far.

* **Test Name:** TestManager_FakeInput_OutputChange
* **Branch:** main
* **Platform** Test / Matrix - PLATFORM = 'windows-2016 && windows-immutable' / Test / TestManager_FakeInput_OutputChange

### Stack Trace

```
Test / Matrix - PLATFORM = 'windows-2016 && windows-immutable' / Test / TestManager_FakeInput_OutputChange – github.com/elastic/elastic-agent/pkg/component/runtime
=== RUN   TestManager_FakeInput_OutputChange
    manager_test.go:2480: component fake-0 state changed: {State:STARTING Message:Starting: spawned pid '2136' Units:map[{UnitType:input UnitID:fake-input-0-0}:{State:STARTING Message:Starting: spawned pid '2136' Payload:map[] unitState:0 unitMessage:Starting unitPayload:map[] configStateIdx:0 err:<nil>} {UnitType:input UnitID:fake-input-0-1}:{State:STARTING Message:Starting: spawned pid '2136' Payload:map[] unitState:0 unitMessage:Starting unitPayload:map[] configStateIdx:0 err:<nil>} {UnitType:input UnitID:fake-input-0-2}:{State:STARTING Message:Starting: spawned pid '2136' Payload:map[] unitState:0 unitMessage:Starting unitPayload:map[] configStateIdx:0 err:<nil>}] VersionInfo:{Name: Version: Meta:map[]} expectedUnits:map[{UnitType:0 UnitID:fake-input-0-0}:{state:2 configStateIdx:1 config:0xc000120a20 err:<nil> logLevel:0} {UnitType:0 UnitID:fake-input-0-1}:{state:2 configStateIdx:1 config:0xc000120ab0 err:<nil> logLevel:0} {UnitType:0 UnitID:fake-input-0-2}:{state:2 configStateIdx:1 config:0xc000120b40 err:<nil> logLevel:0}]}
    manager_test.go:2480: component fake-0 state changed: {State:STOPPED Message:Stopped: pid '2136' exited with code '1' Units:map[{UnitType:input UnitID:fake-input-0-0}:{State:STOPPED Message:Stopped: pid '2136' exited with code '1' Payload:map[] unitState:0 unitMessage:Starting unitPayload:map[] configStateIdx:0 err:<nil>} {UnitType:input UnitID:fake-input-0-1}:{State:STOPPED Message:Stopped: pid '2136' exited with code '1' Payload:map[] unitState:0 unitMessage:Starting unitPayload:map[] configStateIdx:0 err:<nil>} {UnitType:input UnitID:fake-input-0-2}:{State:STOPPED Message:Stopped: pid '2136' exited with code '1' Payload:map[] unitState:0 unitMessage:Starting unitPayload:map[] configStateIdx:0 err:<nil>}] VersionInfo:{Name: Version: Meta:map[]} expectedUnits:map[{UnitType:0 UnitID:fake-input-0-0}:{state:2 configStateIdx:1 config:0xc000120a20 err:<nil> logLevel:0} {UnitType:0 UnitID:fake-input-0-1}:{state:2 configStateIdx:1 config:0xc000120ab0 err:<nil> logLevel:0} {UnitType:0 UnitID:fake-input-0-2}:{state:2 configStateIdx:1 config:0xc000120b40 err:<nil> logLevel:0}]}
    manager_test.go:2484: component fake-1 state changed: {State:STARTING Message:Starting: spawned pid '348' Units:map[{UnitType:input UnitID:fake-input-1-0}:{State:STARTING Message:Starting: spawned pid '348' Payload:map[] unitState:0 unitMessage:Starting unitPayload:map[] configStateIdx:0 err:<nil>} {UnitType:input UnitID:fake-input-1-1}:{State:STARTING Message:Starting: spawned pid '348' Payload:map[] unitState:0 unitMessage:Starting unitPayload:map[] configStateIdx:0 err:<nil>}] VersionInfo:{Name: Version: Meta:map[]} expectedUnits:map[{UnitType:0 UnitID:fake-input-1-0}:{state:2 configStateIdx:1 config:0xc000120c60 err:<nil> logLevel:0} {UnitType:0 UnitID:fake-input-1-1}:{state:2 configStateIdx:1 config:0xc000120cf0 err:<nil> logLevel:0}]}
    manager_test.go:2484: component fake-1 state changed: {State:HEALTHY Message:Healthy: communicating with pid '348' Units:map[{UnitType:input UnitID:fake-input-1-0}:{State:STARTING Message:Starting: spawned pid '348' Payload:map[] unitState:0 unitMessage: unitPayload:map[] configStateIdx:0 err:<nil>} {UnitType:input UnitID:fake-input-1-1}:{State:STARTING Message:Starting: spawned pid '348' Payload:map[] unitState:0 unitMessage: unitPayload:map[] configStateIdx:0 err:<nil>}] VersionInfo:{Name:fake Version:1.0 Meta:map[input:fake]} expectedUnits:map[{UnitType:0 UnitID:fake-input-1-0}:{state:2 configStateIdx:1 config:0xc000120c60 err:<nil> logLevel:0} {UnitType:0 UnitID:fake-input-1-1}:{state:2 configStateIdx:1 config:0xc000120cf0 err:<nil> logLevel:0}]}
    manager_test.go:2484: component fake-1 state changed: {State:HEALTHY Message:Healthy: communicating with pid '348' Units:map[{UnitType:input UnitID:fake-input-1-0}:{State:HEALTHY Message:Fake Healthy 0-0 Payload:map[] unitState:2 unitMessage:Fake Healthy 0-0 unitPayload:map[] configStateIdx:1 err:<nil>} {UnitType:input UnitID:fake-input-1-1}:{State:HEALTHY Message:Fake Healthy 0-1 Payload:map[] unitState:2 unitMessage:Fake Healthy 0-1 unitPayload:map[] configStateIdx:1 err:<nil>}] VersionInfo:{Name:fake Version:1.0 Meta:map[input:fake]} expectedUnits:map[{UnitType:0 UnitID:fake-input-1-0}:{state:2 configStateIdx:1 config:0xc000120c60 err:<nil> logLevel:0} {UnitType:0 UnitID:fake-input-1-1}:{state:2 configStateIdx:1 config:0xc000120cf0 err:<nil> logLevel:0}]}
    manager_test.go:2533: timed out after 30 seconds
--- FAIL: TestManager_FakeInput_OutputChange (34.29s)
```

--- a/pkg/component/fake/component/main.go
+++ b/pkg/component/fake/component/main.go
@@ -68,8 +68,18 @@ func run() error {
 		case <-ctx.Done():
 			return nil
 		case change := <-c.UnitChanges():
-			switch change.Type {
+			if change.Unit != nil {
+				u := change.Unit
+				state, _, _ := u.State()
+				logger.Info().
+					Str("state", state.String()).
+					Str("expectedState", u.Expected().State.String()).
+					Msg("unit change received")
+			} else {
+				logger.Info().Msg("unit change received, but no unit on it")
+			}
 
+			switch change.Type {
 			case client.UnitChangedAdded:
 				s.Added(change.Unit)
 			case client.UnitChangedModified:

--- a/pkg/component/runtime/manager.go
+++ b/pkg/component/runtime/manager.go
@@ -106,6 +106,9 @@ type Manager struct {
 
 	errCh chan error
 
+	// upon creation the Manager is neither running not shutting down, thus both
+	// flags are needed.
+	running      atomic.Bool
 	shuttingDown atomic.Bool
 }
 
@@ -145,6 +148,9 @@ func NewManager(
 //
 // Blocks until the context is done.
 func (m *Manager) Run(ctx context.Context) error {
+	m.running.Store(true)
+	m.shuttingDown.Store(false)
+
 	lis, err := net.Listen("tcp", m.listenAddr)
 	if err != nil {
 		return fmt.Errorf("error starting tcp listener for runtime manager: %w", err)
@@ -202,6 +208,7 @@ func (m *Manager) Run(ctx context.Context) error {
 	}()
 
 	<-ctx.Done()
+	m.running.Store(false)
 	m.shutdown()
 
 	server.Stop()
@@ -286,7 +293,7 @@ func (m *Manager) Errors() <-chan error {
 
 // Update updates the currComp state of the running components.
 //
-// This returns as soon as possible, work is performed in the background to
+// This returns as soon as possible, the work is performed in the background.
 func (m *Manager) Update(components []component.Component) error {
 	shuttingDown := m.shuttingDown.Load()
 	if shuttingDown {

--- a/pkg/component/runtime/manager_test.go
+++ b/pkg/component/runtime/manager_test.go
@@ -13,6 +13,7 @@ import (
 	"path/filepath"
 	"regexp"
 	"runtime"
+	"strings"
 	"testing"
 	"time"
 
@@ -1941,9 +1942,17 @@ func TestManager_FakeInput_MultiComponent(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
-	ai, _ := info.NewAgentInfo(true)
-	m, err := NewManager(newDebugLogger(t), newDebugLogger(t), "localhost:0", ai, apmtest.DiscardTracer, newTestMonitoringMgr(), configuration.DefaultGRPCConfig())
+	agentInfo, _ := info.NewAgentInfo(true)
+	m, err := NewManager(
+		newDebugLogger(t),
+		newDebugLogger(t),
+		"localhost:0",
+		agentInfo,
+		apmtest.DiscardTracer,
+		newTestMonitoringMgr(),
+		configuration.DefaultGRPCConfig())
 	require.NoError(t, err)
+
 	errCh := make(chan error)
 	go func() {
 		err := m.Run(ctx)
@@ -2606,16 +2615,24 @@ LOOP:
 }
 
 func TestManager_FakeInput_OutputChange(t *testing.T) {
-	t.Skip("Flaky test: https://github.com/elastic/elastic-agent/issues/2403")
 	testPaths(t)
 
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
 	ai, _ := info.NewAgentInfo(true)
-	m, err := NewManager(newDebugLogger(t), newDebugLogger(t), "localhost:0", ai, apmtest.DiscardTracer, newTestMonitoringMgr(), configuration.DefaultGRPCConfig())
-	require.NoError(t, err)
+	m, err := NewManager(
+		newDebugLogger(t),
+		newDebugLogger(t),
+		"localhost:0",
+		ai,
+		apmtest.DiscardTracer,
+		newTestMonitoringMgr(),
+		configuration.DefaultGRPCConfig())
+	require.NoError(t, err, "could not crete new manager")
+
 	errCh := make(chan error)
+	t.Cleanup(func() { drainErrChan(errCh) })
 	go func() {
 		err := m.Run(ctx)
 		if errors.Is(err, context.Canceled) {
@@ -2637,9 +2654,12 @@ func TestManager_FakeInput_OutputChange(t *testing.T) {
 		BinaryPath: binaryPath,
 		Spec:       fakeInputSpec,
 	}
+	const IDComp0 = "fake-0"
+	const IDComp1 = "fake-1"
+
 	components := []component.Component{
 		{
-			ID:        "fake-0",
+			ID:        IDComp0,
 			InputSpec: &runtimeSpec,
 			Units: []component.Unit{
 				{
@@ -2675,7 +2695,7 @@ func TestManager_FakeInput_OutputChange(t *testing.T) {
 
 	components2 := []component.Component{
 		{
-			ID:        "fake-1",
+			ID:        IDComp1,
 			InputSpec: &runtimeSpec,
 			Units: []component.Unit{
 				{
@@ -2713,85 +2733,97 @@ func TestManager_FakeInput_OutputChange(t *testing.T) {
 		componentID string
 		state       ComponentState
 	}
-	stateProgression := make([]progressionStep, 0)
+	var stateProgression []progressionStep
 
 	subCtx, subCancel := context.WithCancel(context.Background())
 	defer subCancel()
-	stateProgCh := make(chan progressionStep)
+
+	stateProgressionCh := make(chan progressionStep)
 	subErrCh0 := make(chan error)
 	subErrCh1 := make(chan error)
+	t.Cleanup(func() { drainErrChan(subErrCh0) })
+	t.Cleanup(func() { drainErrChan(subErrCh1) })
+
 	go func() {
-		sub0 := m.Subscribe(subCtx, "fake-0")
-		sub1 := m.Subscribe(subCtx, "fake-1")
+		sub0 := m.Subscribe(subCtx, IDComp0)
+		sub1 := m.Subscribe(subCtx, IDComp1)
 		for {
 			select {
 			case <-subCtx.Done():
-				close(stateProgCh)
+				close(stateProgressionCh)
 				return
 			case state := <-sub0.Ch():
-				t.Logf("component fake-0 state changed: %+v", state)
-				signalState(subErrCh0, &state, []client.UnitState{client.UnitStateHealthy, client.UnitStateStopped})
-				stateProgCh <- progressionStep{"fake-o", state}
+				t.Logf("component %s state changed: %+v", IDComp0, state)
+				signalState(
+					subErrCh0,
+					&state,
+					[]client.UnitState{client.UnitStateHealthy, client.UnitStateStopped})
+				stateProgressionCh <- progressionStep{IDComp0, state}
+
 			case state := <-sub1.Ch():
-				t.Logf("component fake-1 state changed: %+v", state)
-				signalState(subErrCh1, &state, []client.UnitState{client.UnitStateHealthy})
-				stateProgCh <- progressionStep{"fake-o", state}
+				t.Logf("component %s state changed: %+v", IDComp1, state)
+				signalState(
+					subErrCh1,
+					&state,
+					[]client.UnitState{client.UnitStateHealthy})
+				stateProgressionCh <- progressionStep{IDComp1, state}
 			}
 		}
 	}()
 
 	go func() {
-		for step := range stateProgCh {
+		for step := range stateProgressionCh {
 			stateProgression = append(stateProgression, step)
 		}
 	}()
 
-	defer drainErrChan(errCh)
-	defer drainErrChan(subErrCh0)
-	defer drainErrChan(subErrCh1)
+	// Wait manager start running, then check if any error happened
+	assert.Eventually(t,
+		func() bool { return m.running.Load() },
+		500*time.Millisecond,
+		10*time.Millisecond)
 
-	startTimer := time.NewTimer(100 * time.Millisecond)
-	defer startTimer.Stop()
 	select {
-	case <-startTimer.C:
-		err = m.Update(components)
-		require.NoError(t, err)
 	case err := <-errCh:
 		t.Fatalf("failed early: %s", err)
+	default:
 	}
 
-	updateTimeout := 300 * time.Millisecond
+	time.Sleep(100 * time.Millisecond)
+	err = m.Update(components)
+	require.NoError(t, err)
+
+	updateSleep := 300 * time.Millisecond
 	if runtime.GOOS == component.Windows {
-		// windows is slow, preventing flakyness
-		updateTimeout = 550 * time.Millisecond
+		// windows is slow, preventing flakiness
+		updateSleep = time.Second
 	}
-	updateTimer := time.NewTimer(updateTimeout)
-	defer updateTimer.Stop()
-	select {
-	case <-updateTimer.C:
-		err = m.Update(components2)
-		require.NoError(t, err)
-	case err := <-errCh:
-		t.Fatalf("failed early: %s", err)
-	}
+	time.Sleep(updateSleep)
+	err = m.Update(components2)
+	require.NoError(t, err)
 
 	count := 0
-	endTimer := time.NewTimer(30 * time.Second)
+	timeout := 30 * time.Second
+	endTimer := time.NewTimer(timeout)
 	defer endTimer.Stop()
+
 LOOP:
 	for {
 		select {
 		case <-endTimer.C:
-			t.Fatalf("timed out after 30 seconds")
+			t.Fatalf("timed out after %s seconds, "+
+				"did not receive enought state changes", timeout)
 		case err := <-errCh:
 			require.NoError(t, err)
 		case err := <-subErrCh0:
+			t.Logf("[subErrCh0] received: %v", err)
 			require.NoError(t, err)
 			count++
 			if count >= 2 {
 				break LOOP
 			}
 		case err := <-subErrCh1:
+			t.Logf("[subErrCh1] received: %v", err)
 			require.NoError(t, err)
 			count++
 			if count >= 2 {
@@ -2803,14 +2835,16 @@ LOOP:
 	subCancel()
 	cancel()
 
-	// check progresstion, require stop fake-0 before start fake-1
-	wasStopped := false
+	// check progression, require stop fake-0 before start fake-1
+	comp0Stopped := false
 	for _, step := range stateProgression {
-		if step.componentID == "fake-0" && step.state.State == client.UnitStateStopped {
-			wasStopped = true
+		if step.componentID == IDComp0 &&
+			step.state.State == client.UnitStateStopped {
+			comp0Stopped = true
 		}
-		if step.componentID == "fake-1" && step.state.State == client.UnitStateStarting {
-			require.True(t, wasStopped)
+		if step.componentID == IDComp1 &&
+			step.state.State == client.UnitStateStarting {
+			require.True(t, comp0Stopped)
 			break
 		}
 	}
@@ -2844,28 +2878,34 @@ func drainErrChan(ch chan error) {
 func signalState(subErrCh chan error, state *ComponentState, acceptableStates []client.UnitState) {
 	if state.State == client.UnitStateFailed {
 		subErrCh <- fmt.Errorf("component failed: %s", state.Message)
-	} else {
-		issue := ""
-		healthy := 0
-		for key, unit := range state.Units {
-			if unit.State == client.UnitStateStarting {
-				// acceptable
-			} else if isAcceptableState(unit.State, acceptableStates) {
-				healthy++
-			} else if issue == "" {
-				issue = fmt.Sprintf("unit %s in invalid state %v", key.UnitID, unit.State)
-			}
+		return
+	}
+
+	var issues []string
+	healthy := 0
+	for key, unit := range state.Units {
+		switch {
+		case unit.State == client.UnitStateStarting:
+		// acceptable, but does not count as health
+		case isValidState(unit.State, acceptableStates):
+			healthy++
+		default:
+
+			issues = append(issues, fmt.Sprintf(
+				"unit %s in invalid state %v", key.UnitID, unit.State))
 		}
-		if issue != "" {
-			subErrCh <- fmt.Errorf("%s", issue)
-		}
-		if healthy == 3 {
-			subErrCh <- nil
-		}
+	}
+
+	if len(issues) != 0 {
+		subErrCh <- errors.New(strings.Join(issues, "| "))
+	}
+
+	if healthy == len(state.Units) {
+		subErrCh <- nil
 	}
 }
 
-func isAcceptableState(state client.UnitState, acceptableStates []client.UnitState) bool {
+func isValidState(state client.UnitState, acceptableStates []client.UnitState) bool {
 	for _, s := range acceptableStates {
 		if s == state {
 			return true

